library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
Note that the `alternative` laws are upheld for non-backtracking parsers, which I believe are better for error reporting, and also what the current `interactive` parser is doing (and what `tactic` should also be doing, imo). I used non-backtracking combinators for the JSON parser for POPL that I never finished: https://gist.github.com/Kha/9a9a19ee1dd3331829f182555dddaf0b#file-parser_combinator-lean-L53